release/release-offline.ts

```diff
@@ -0,0 +1,410 @@
+ * you should only use this script if absolutely necessary, and you cannot use CI for some very
+ * good reason.
+ */
+
+import "dotenv/config";
+import { Octokit } from "@octokit/rest";
+import "zx/globals";
+import { $ } from "zx";
+$.verbose = false;
+
+import {
+  isValidVersionString,
+  hasBeenReleased,
+  isValidCommitHash,
+  isEnterpriseVersion,
+  getMajorVersion,
+  isLatestRelease,
+  getVersionInfo,
+  publishRelease,
+  closeMilestone,
+  openNextMilestones,
+  versionRequirements,
+} from "./src";
+
+const {
+  GITHUB_TOKEN,
+  GITHUB_OWNER,
+  GITHUB_REPO,
+  AWS_S3_DOWNLOADS_BUCKET,
+  AWS_S3_STATIC_BUCKET,
+  AWS_CLOUDFRONT_DOWNLOADS_ID,
+  AWS_CLOUDFRONT_STATIC_ID,
+  AWS_S3_RELEASE_SECRET_ACCESS_KEY,
+  AWS_S3_RELEASE_ACCESS_KEY_ID,
+  DOCKERHUB_OWNER,
+  DOCKERHUB_RELEASE_USERNAME,
+  DOCKERHUB_RELEASE_TOKEN,
+} = process.env;
+
+const github = new Octokit({ auth: GITHUB_TOKEN });
+
+const JAR_PATH = "../target/uberjar";
+
+const version = process.argv?.[2]?.trim();
+const commit = process.argv?.[3]?.trim();
+const step = process.argv?.[4]?.trim().replace("--", "");
+const edition = isEnterpriseVersion(version) ? "ee" : "oss";
+
+if (!isValidVersionString(version)) {
+  error(
+    "You must provide a valid version string as the first argument (e.g v0.45.6",
+  );
+}
+
+if (!isValidCommitHash(commit)) {
+  error("You must provide a valid commit hash as the second argument");
+}
+
+if (!step) {
+  error("You must provide a step argument like --build or --publish");
+}
+
+/**************************************************
+          HELPERS
+ **************************************************/
+
+// mostly for type checking
+function getGithubCredentials() {
+  if (!GITHUB_TOKEN || !GITHUB_OWNER || !GITHUB_REPO) {
+    error("You must provide all github environment variables in .env-template");
+    process.exit(1);
+  }
+
+  return { GITHUB_TOKEN, GITHUB_OWNER, GITHUB_REPO };
+}
+
+async function checkReleased() {
+  const { GITHUB_OWNER, GITHUB_REPO } = getGithubCredentials();
+
+  const hasThisVersionBeenReleased = await hasBeenReleased({
+    github,
+    owner: GITHUB_OWNER,
+    repo: GITHUB_REPO,
+    version,
+  });
+
+  if (hasThisVersionBeenReleased) {
+    error(`Version ${version} has already been released`);
+  }
+}
+
+async function checkJar() {
+  // check if built jar exists
+  const requiredFiles = [
+    `${JAR_PATH}/metabase.jar`,
+    `${JAR_PATH}/COMMIT-ID`,
+    `${JAR_PATH}/SHA256.sum`,
+  ];
+
+  requiredFiles.forEach(file => {
+    if (!fs.existsSync(file)) {
+      error(`You must build the jar first. ${file} does not exist`);
+    }
+  });
+
+  const versionProperties = (
+    await $`jar xf ${JAR_PATH}/metabase.jar version.properties && cat version.properties`
+  ).toString();
+
+  log(versionProperties, "green");
+
+  if (!versionProperties.includes(`tag=${version}`)) {
+    error(`This jar does not match the input version: ${version}`);
+  }
+}
+
+// eslint-disable-next-line no-console
```

Comment:
This comment needs to be in the next line

Reply:
prettier did me dirty on this one